AK: Make BumpAllocator work in multi-threaded environments

BenWiederhake opened the pull request:
Fixes #10578.

I also implemented setting the pages to volatile while they're in the cache. However, this increases the running time of the `alternative_match_groups_benchmark` from about 2.1 seconds to "I got tired of waiting after several minutes". So I didn't make it part of this commit :D

alimpfard commented on AK/BumpAllocator.h:
```diff
@@ -55,14 +56,22 @@ class BumpAllocator {
+        // then we haven't lost much; it was a close call anyway.
+        // If we determine 'cache_filled=false' and the cache becomes full in the meantime,
+        // then we'll end up with a different chunk to munmap(), no big difference.
+        bool cache_filled = s_unused_allocation_cache.load();
```
The default memory order is quite overkill for this, this initial load can be `memory_order_relaxed`.
The exchanges can also be relaxed a bit, but we can afford overkill on that one since it's a much more rare event than deallocation.

BenWiederhake replied:
I don't really know too much about the various memory ordering guarantees; that's what the class is for, after all.

Either way, I didn't see any significant or even measurable difference. (Seriously, the times taken for the benchmark were all over the place, so comparing them didn't give any result.) Therefore, I'd prefer to stick to the default memory ordering. I'll change this to "relaxed" though, since the load really is only a nice-to-have heuristic, and in case of contention it doesn't matter which value we get.

Exchanges happen basically twice as often as the load here (once during allocation, once during deallocation). So I'm not entirely sure what you mean, but I guess this will be good enough.

Reading through the [reference](https://en.cppreference.com/w/cpp/atomic/memory_order), I feel like the exchanges can be optimized like this:
- `memory_order_acquire` for the exchange in `allocate_a_chunk()` (to ensure that the `new_header.magic = chunk_magic` and following won't happen too early)
- `memory_order_release` for the exchange in `deallocate_all()` (to ensure that the `((ChunkHeader*)chunk)->next_chunk = 0` is done; if this ever turns out to be time-critical we can probably delete the overwrite and use `relaxed`, since `allocate_a_chunk()` already clears that field.)
- However, I'm not confident enough in these hyper-optimizations, and they don't seem to make a difference anyway, so I won't do them.

alimpfard replied:
Yeah, I'm just suggesting to make the load use relaxed, the exchanges can be dealt with later if they become a bottleneck.

alimpfard reviewed (commented):
Just a little nit, otherwise LGTM

BenWiederhake commented:
Changes since last push:
- Rebased
- The `cache_filled` heuristic now uses the "relaxed" memory order, since it only need to roughly correlate with reality, and if there is any contention then the result doesn't actually matter.
